spec/typecheck/typecheck_error_spec.lua

```diff
@@ -0,0 +1,73 @@
+local tl = require("tl")
+
+describe("typecheck errors", function()
+   it("type errors include filename", function ()
+      local tokens = tl.lex("local x: string = 1")
+      local _, ast = tl.parse_program(tokens, {})
+      local errors, unknowns = tl.type_check(ast, false, "foo.tl")
+      assert.same("foo.tl", errors[1].filename, "type errors should contain .filename property")
+   end)
+
+   it("type errors in a required package include filename of required file", function ()
+      local io_open = io.open
+      io.open = function (filename, mode)
+         if string.match(filename, "bar.tl$") then
+            -- Return a stub file handle
+            return {
+               read = function (_, format)
+                  if format == "*a" then
+                     return "local x: string = 1"        -- Return fake bar.tl content
+                  else
+                     error("Not implemented!")  -- Implement other modes if needed
+                  end
+               end,
+               close = function () end,
+            }
+         else
+            return io_open(filename, mode)
+         end
+      end
+
+      local tokens = tl.lex([[
+         local bar = require "bar"
+      ]])
+      local _, ast = tl.parse_program(tokens, {}, "foo.tl")
+      local errors, unknowns = tl.type_check(ast, true, "foo.tl")
+      assert.is_not_nil(string.match(errors[1].filename, "bar.tl$"), "type errors should contain .filename property")
+   end)
+
+   it("unknowns include filename", function ()
+      local tokens = tl.lex("local x: string = b")
+      local _, ast = tl.parse_program(tokens, {})
+      local errors, unknowns = tl.type_check(ast, true, "foo.tl")
+      assert.same("foo.tl", unknowns[1].filename, "unknowns should contain .filename property")
+   end)
+
+   it("unknowns in a required package include filename of required file", function ()
+      local io_open = io.open
+      io.open = function (filename, mode)
```

Comment:
when you monkey-patch a function like this in a test, I think it's necessary to switch it back once the test is done, otherwise if a new test is added later to the file that assumes the function is the standard one, it will fail.

You can use a `finally(function() io.open = io_open end)` block to ensure the function is reverted to the original on both test success and failure

Reply:
Good point, thanks, i'll fix this.

There's additional test sandboxing support in busted, called "insulation" [1], but that wouldn't help either in this case looks like it - it only sandboxes loaded packages `package.loaded` and globals, and in this case it is _modifying_ a package.

[1] https://olivinelabs.com/busted/#insulate-expose-blocks